src/django_components/node.py

```diff
@@ -1,20 +1,460 @@
+            for resolved_param in resolved_params:
+                key = resolved_param.key
+                if key is not None and (not key.isidentifier() or keyword.iskeyword(key)):
+                    invalid_kwargs[key] = resolved_param.value
+                else:
+                    resolved_params_without_invalid_kwargs.append(resolved_param)
+
+            # Validate the params against the signature
+            #
+            # Unlike the call to `apply_params_in_original_order()` further below, this uses a signature
+            # that has been stripped of the `self` and `context` parameters. E.g.
+            #
+            # `def render(name: str, **kwargs: Any) -> None`
+            #
+            # If there are any errors in the input, this will trigger Python's
+            # native error handling (e.g. `TypeError: render() got multiple values for argument 'context'`)
+            #
+            # But because we stripped the two parameters, then these errors will correctly
+            # point to the actual error in the template tag.
+            #
+            # E.g. if we supplied one too many positional args,
+            # `{% mytag "John" 20 %}`
+            #
+            # Then without stripping the two parameters, then the error could be:
+            # `render() takes from 3 positional arguments but 4 were given`
+            #
+            # Which is confusing, because we supplied only two positional args.
+            #
+            # But cause we stripped the two parameters, then the error will be:
+            # `render() takes from 1 positional arguments but 2 were given`
+            validate_params(self.tag, validation_signature, resolved_params_without_invalid_kwargs, invalid_kwargs)
+
+            # The code below calls the `orig_render()` function like so:
+            # `orig_render(self, context, arg1, arg2, kwarg1=val1, kwarg2=val2)`
+            #
+            # So it's called in the same order as what was passed to the template tag, e.g.
+            # `{% component arg1 arg2 kwarg1=val1 kwarg2=val2 %}`
+            #
+            # That's why we don't simply spread all args and kwargs as `*args, **kwargs`,
+            # because then Python's validation wouldn't catch such errors.
+            resolved_params_with_context = [
+                TagParam(key=None, value=self),
+                TagParam(key=None, value=context),
+            ] + resolved_params_without_invalid_kwargs
+            output = apply_params_in_original_order(orig_render, resolved_params_with_context, invalid_kwargs)
+
+            trace_msg("RENDR", self.tag, self.node_id, msg="...Done!")
+            return output
+
+        # Wrap cls.render() so we resolve the args and kwargs and pass them to the
+        # actual render method.
+        cls.render = wrapper_render  # type: ignore
+        cls.render._djc_wrapped = True  # type: ignore
+
+        return cls
+
+
+class BaseNode(Node, metaclass=NodeMeta):
```

Comment:
In this file is the core of this PR, which adds the `BaseNode` class and `@template_tag` decorator, for an easy creation of template tags that use the same features like other django-components' tags.